Make DynamoDB clients injectable beans

marcinczeczko opened the pull request:
This is a continuation of #2193 - DynamoDB as injectable beans, configured via application.properties config

marcinczeczko committed:
DynamoDB as injectable bean

marcinczeczko commented:
@gsmet - please take a look at my approach to the dynamo bean. It's still work in progress, but maybe you could have some suggestions.

gsmet reviewed (changes requested):
@marcinczeczko interesting work, thanks!

I added some comment. I think we will need to generate an example config file and check the consistency of the config, once we are done with the rest.

The main issue is about trying to only create the required clients instead of always creating both. @mkouba had some nice proposals about this so I dumped them in the comments.

Feel free to ask questions or comment the approach if you have doubts.

gsmet commented:
Hey @marcinczeczko 

Do you still plan to make progress on this?

marcinczeczko commented:
I @gsmet sorry for no replies. I literally just came back from holidays. I'll update you tomorrow what are the next steps. Anyway, thanks for your comments especially regards how to avoid initialization of both clients if one is used - this is something I was looking for so I'd go towards that direction.

gsmet commented:
@marcinczeczko OK cool, I hope you had a nice time away from keyboards :).

marcinczeczko commented on extensions/amazon-dynamodb/runtime/src/main/java/io/quarkus/dynamodb/runtime/AwsNettyNioAsyncHttpClientConfig.java:
```diff
@@ -0,0 +1,91 @@
+     * SSL Provider
+     */
+    @ConfigItem
+    public Optional<SslProvider> sslProvider;
```
What do you mean ? SslProvider is just enum.

gsmet replied:
Ah totally missed that.

marcinczeczko commented on extensions/amazon-dynamodb/deployment/src/main/java/io/quarkus/dynamodb/deployment/DynamodbProcessor.java:
```diff
@@ -1,34 +1,58 @@
         // Indicates that this extension would like the SSL support to be enabled
         extensionSslNativeSupport.produce(new ExtensionSslNativeSupportBuildItem(FeatureBuildItem.DYNAMODB));
+
+        additionalBeans.produce(AdditionalBeanBuildItem.unremovableOf(DynamodbClientProducer.class));
```
Fixed.

marcinczeczko commented on extensions/amazon-dynamodb/runtime/src/main/java/io/quarkus/dynamodb/runtime/AwsCredentialsProviderConfig.java:
```diff
@@ -0,0 +1,132 @@
+    /**
+     * Credentials provider
+     */
+    @ConfigItem(name = "type", defaultValue = "DEFAULT")
```
Fixed

marcinczeczko commented on extensions/amazon-dynamodb/runtime/src/main/java/io/quarkus/dynamodb/runtime/AwsApacheHttpClientConfig.java:
```diff
@@ -0,0 +1,120 @@
+     * HTTP Proxy configuration
+     */
+    @ConfigItem
+    HttpClientProxyConfiguration proxyConfiguration;
```
FIxed

marcinczeczko commented on extensions/amazon-dynamodb/runtime/src/main/java/io/quarkus/dynamodb/runtime/AwsApacheHttpClientConfig.java:
```diff
@@ -0,0 +1,120 @@
+     * Enabled http proxy
+     */
+    @ConfigItem(defaultValue = "false")
+    public boolean proxyEnabled;
```
Good point. Fixed

marcinczeczko commented on extensions/amazon-dynamodb/runtime/src/main/java/io/quarkus/dynamodb/runtime/AwsApacheHttpClientConfig.java:
```diff
@@ -0,0 +1,120 @@
+         * Defines set of non-proxy hosts
+         */
+        @ConfigItem
+        public Optional<Set<String>> nonProxyHosts;
```
Indeed. Fixed.

marcinczeczko commented on extensions/amazon-dynamodb/runtime/src/main/java/io/quarkus/dynamodb/runtime/DynamodbClientConfig.java:
```diff
@@ -0,0 +1,54 @@
+     * Apache HTTP Client transport configuration
+     */
+    @ConfigItem
+    public AwsApacheHttpClientConfig apacheHttpClientConfig;
```
Agree. it will shorten keys in the config. Fixed

marcinczeczko commented on extensions/amazon-dynamodb/runtime/src/main/java/io/quarkus/dynamodb/runtime/DynamodbClientConfig.java:
```diff
@@ -0,0 +1,54 @@
+     * Netty HTTP Client transport configuration
+     */
+    @ConfigItem
+    public AwsNettyNioAsyncHttpClientConfig nettyHttpClientConfig;
```
Fixed

marcinczeczko commented on extensions/amazon-dynamodb/runtime/src/main/java/io/quarkus/dynamodb/runtime/DynamodbClientConfig.java:
```diff
@@ -0,0 +1,54 @@
+     * Defines credentials provider used
+     */
+    @ConfigItem
+    public AwsCredentialsProviderConfig credentialsConfig;
```
Good point. Fixed

marcinczeczko commented:
Hi @gsmet - I just pushed first version of the changes regarding lazy creation of clients. It's not a ready work - some refactoring and cleanup is needed. I'll appreciate any advise or comments if I'm heading in right direction. Thanks

marcinczeczko commented on integration-tests/amazon-dynamodb/src/main/resources/application.properties:
```diff
@@ -1,3 +1,10 @@
+quarkus.dynamodb.region=localhost
+quarkus.dynamodb.endpoint-override=${dynamodb.url}
+
+quarkus.dynamodb.credentials-config.type=STATIC
```
fixed

marcinczeczko commented on extensions/amazon-dynamodb/runtime/src/main/java/io/quarkus/dynamodb/runtime/DynamodbRecorder.java:
```diff
@@ -0,0 +1,142 @@
+            BeanContainer container, ShutdownContext shutdown) {
+
+        initClient(config);
+        initAsyncClient(config);
```
Yes. See my latest commit. However still need your advise/comment if I'm going into right direction.

marcinczeczko commented:
Hi @gsmet - did you have chance to look at the latest changes ?

gsmet commented:
No, not yet, it's on my list for the next few days. I'll get back to you soon.

gsmet commented:
@marcinczeczko you have a formatting error, please run `mvn process-sources` and then amend the related commit.

FWIW, I'll start reviewing this now (and will probably post the review tomorrow morning).

marcinczeczko commented:
@gsmet - I fixed formatting.

gsmet commented on extensions/amazon-dynamodb/runtime/src/main/java/io/quarkus/dynamodb/runtime/AwsApacheHttpClientConfig.java:
```diff
@@ -0,0 +1,119 @@
+public class AwsApacheHttpClientConfig {
+
+    /**
+     * Connection Acquisition Timeout
```
We usually don't all capitalize the javadoc. It should be `Connection acquisition timeout`.

Granted it will be very tedious to change that so not a blocker for me.

marcinczeczko replied:
I de-captilized it anyway :)

gsmet reviewed (changes requested):
@marcinczeczko I think the approach is generally OK, nice work.

I added a couple of comments here and there.

marcinczeczko commented on extensions/amazon-dynamodb/runtime/src/main/java/io/quarkus/dynamodb/runtime/DynamodbClientProducer.java:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    @Produces
+    public DynamoDbClient client() {
```
I'm using `@ApplicationScope` on the producer class level. Is it required on the method level ? Sorry, I'm trying to understand the difference

gsmet replied:
Well, defining it at the producer level defines the scope of the producer itself. Then you want to define the scope of the bean.

marcinczeczko commented on integration-tests/amazon-dynamodb/src/main/resources/application.properties:
```diff
@@ -1,3 +1,10 @@
 quarkus.log.category.\"software.amazon.awssdk.request\".level=DEBUG
-quarkus.log.category.\"org.apache.http.wire\".level=DEBUG
\ No newline at end of file
+quarkus.log.category.\"org.apache.http.wire\".level=DEBUG
```
Yep, I'll clean it up

marcinczeczko commented on extensions/amazon-dynamodb/runtime/src/main/java/io/quarkus/dynamodb/runtime/DynamodbClientConfig.java:
```diff
@@ -0,0 +1,55 @@
+ *
+ */
+@ConfigRoot(name = "dynamodb", phase = ConfigPhase.RUN_TIME)
+public class DynamodbClientConfig {
```
Done

marcinczeczko commented on extensions/amazon-dynamodb/runtime/src/main/java/io/quarkus/dynamodb/runtime/AwsApacheHttpClientConfig.java:
```diff
@@ -0,0 +1,119 @@
+        public Optional<String> ntlmDomain;
+
+        /**
+         * HTML Workstation
```
Done

marcinczeczko commented on extensions/amazon-dynamodb/deployment/src/main/java/io/quarkus/dynamodb/deployment/DynamodbProcessor.java:
```diff
@@ -1,58 +1,134 @@
 import software.amazon.awssdk.http.nio.netty.NettySdkAsyncHttpService;
+import software.amazon.awssdk.services.dynamodb.DynamoDbAsyncClient;
+import software.amazon.awssdk.services.dynamodb.DynamoDbClient;
+import software.amazon.awssdk.utils.StringUtils;
```
I was thinking about this. I'm using `isBlank` & `isNotBlank` util functions - wanted to avoid re-implementing it in here. Maybe it's worth to supply apache commons with all those utils at the quarkus bom level ? wdyt ?

gsmet replied:
We try to avoid including too many external dependencies. Let's leave it at that for now.

marcinczeczko commented on extensions/amazon-dynamodb/deployment/src/main/java/io/quarkus/dynamodb/deployment/DynamodbProcessor.java:
```diff
@@ -1,58 +1,134 @@
+            if (SYNC_CLIENT_NAME.equals(requiredType.name()) || ASYNC_CLIENT_NAME.equals(requiredType.name())) {
+                clientBeanTypes.add(requiredType);
+            } else {
+                continue;
```
Thanks for nice suggestion. I fixed it

marcinczeczko commented on extensions/amazon-dynamodb/deployment/src/main/java/io/quarkus/dynamodb/deployment/DynamodbProcessor.java:
```diff
@@ -1,58 +1,134 @@
+                        new ServiceProviderBuildItem(SdkAsyncHttpService.class.getName(),
+                                NettySdkAsyncHttpService.class.getName()));
+
+                clients.produce(new DynamodbClientBuildItem(false));
```
Fixed

marcinczeczko commented on extensions/amazon-dynamodb/deployment/src/main/java/io/quarkus/dynamodb/deployment/DynamodbProcessor.java:
```diff
@@ -1,58 +1,134 @@
-        //Register netty as async client
-        serviceProvider.produce(
-                new ServiceProviderBuildItem(SdkAsyncHttpService.class.getName(), NettySdkAsyncHttpService.class.getName()));
+        for (DynamodbClientBuildItem client : clientItems) {
```
Indeed. Fixed.

marcinczeczko commented:
@gsmet  - I squashed all my commits into one and rebased to have clean history. Already created quickstart for dynamodb https://github.com/quarkusio/quarkus-quickstarts/pull/261 - to be updated once this PR is released.

marcinczeczko commented:
Added guide under #3706

gsmet commented on extensions/amazon-dynamodb/deployment/src/main/java/io/quarkus/dynamodb/deployment/DynamodbProcessor.java:
```diff
@@ -1,58 +1,208 @@
-    void setupDynamoDb() {
+    SubstrateProxyDefinitionBuildItem httpProxies() {
+        return new SubstrateProxyDefinitionBuildItem("org.apache.http.conn.HttpClientConnectionManager",
+                "org.apache.http.pool.ConnPoolControl", "software.amazon.awssdk.http.apache.internal.conn.Wrapped");
```
Are they useful in both cases? Looks related to the sync client only? If so maybe we should condition them?

marcinczeczko replied:
Fixed

gsmet commented on extensions/amazon-dynamodb/deployment/src/main/java/io/quarkus/dynamodb/deployment/DynamodbProcessor.java:
```diff
@@ -1,58 +1,208 @@
+    DynamodbClientBuildItem analyzeDynamodbClientInjectionPoints(BeanRegistrationPhaseBuildItem beanRegistrationPhase,
+            BuildProducer<ServiceProviderBuildItem> serviceProvider) {
+
+        boolean createSyncClient = false, createAsyncClient = false;
```
Really not a big fan of this. Please one per line.

marcinczeczko replied:
Fixed

gsmet commented on extensions/amazon-dynamodb/runtime/src/main/java/io/quarkus/dynamodb/runtime/AwsCredentialsProviderConfig.java:
```diff
@@ -0,0 +1,132 @@
+            return new StringJoiner(", ", StaticCredentialsProviderConfig.class.getSimpleName() + "[", "]")
+                    .add("accessKeyId='" + accessKeyId + "'")
+                    .add("secretAccessKey='" + secretAccessKey + "'")
+                    .toString();
```
Any chance you could take this remark into account?

marcinczeczko replied:
I removed toString() anyway as seems to be a leftover

gsmet commented on extensions/amazon-dynamodb/runtime/src/main/java/io/quarkus/dynamodb/runtime/AwsCredentialsProviderConfig.java:
```diff
@@ -0,0 +1,132 @@
+         *
+         * <p>
+         * By default, this is disabled.
+         * </p>
```
This remark got lost apparently :).

marcinczeczko replied:
Fixed. `</p>` in javadoc.

gsmet reviewed (changes requested):
I added a couple more comments. I think we're near the end for this one.

BTW, unfortunately, there is a conflict with the new pipeline stuff, you will need to rebase and port your changes.

marcinczeczko commented on extensions/amazon-dynamodb/runtime/src/main/java/io/quarkus/dynamodb/runtime/AwsApacheHttpClientConfig.java:
```diff
@@ -0,0 +1,119 @@
+    @ConfigGroup
+    public static class HttpClientProxyConfiguration {
+        /**
+         * Enabled http proxy
```
Fixed

marcinczeczko commented on extensions/amazon-dynamodb/runtime/src/main/java/io/quarkus/dynamodb/runtime/AwsApacheHttpClientConfig.java:
```diff
@@ -0,0 +1,119 @@
+        public Optional<String> ntlmWorkstation;
+
+        /**
+         * Enabled Preemptive Basic Authentication
```
fixed

marcinczeczko commented on extensions/amazon-dynamodb/runtime/src/main/java/io/quarkus/dynamodb/runtime/AwsApacheHttpClientConfig.java:
```diff
@@ -0,0 +1,119 @@
+         * Defines list of non-proxy hosts
+         */
+        @ConfigItem
+        public List<String> nonProxyHosts;
```
Well, I used the same wording as in the AWS SDK which is `nonProxyHosts` - anyway updated javadoc to be more verbose

marcinczeczko commented on extensions/amazon-dynamodb/runtime/src/main/java/io/quarkus/dynamodb/runtime/AwsCredentialsProviderConfig.java:
```diff
@@ -0,0 +1,132 @@
+        public Optional<Boolean> asyncCredentialUpdateEnabled;
+
+        /**
+         * Reuse last provided
```
Javadoc updated

marcinczeczko commented on extensions/amazon-dynamodb/runtime/src/main/java/io/quarkus/dynamodb/runtime/DynamodbClientConfig.java:
```diff
@@ -0,0 +1,55 @@
+import software.amazon.awssdk.regions.Region;
+
+/**
+ * @author Marcin Czeczko
```
Removed

marcinczeczko commented on extensions/amazon-dynamodb/runtime/src/main/java/io/quarkus/dynamodb/runtime/DynamodbClientProducer.java:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    @Produces
+    public DynamoDbAsyncClient asyncClient() {
```
Added

marcinczeczko commented:
@gsmet - Review fixed done. Sorry I missed other remarks before - they were hidden because of forced push :P

gsmet commented:
@marcinczeczko sorry there is apparently a conflict again (we are making some changes to the CI infrastructure). Could you have another look?

It would be nice if we could get this one in today.

marcinczeczko commented:
@gsmet - done

marcinczeczko commented:
Looks like native integration tests of `Hibernate ORM with Panache` are failing.

gsmet reviewed (approved):
Let's wait for CI.

gsmet commented:
Merged, thanks for your work on this!
